src/audio/google/google_rtc_audio_processing.c

```diff
@@ -639,6 +734,190 @@ static int google_rtc_audio_processing_reset(struct processing_module *mod)
 	return 0;
 }
 
+#if CONFIG_IPC_MAJOR_4
+#if CONFIG_COMP_GOOGLE_RTC_USE_32_BIT_FLOAT_API
+static inline int16_t convert_float_to_int16(float data)
+{
+#if XCHAL_HAVE_HIFI3
+	const xtfloat ratio = 2 << 15;
+	xtfloat x0 = data;
+	xtfloat x1;
+	int16_t x;
+
+	x1 = XT_MUL_S(x0, ratio);
+	x = XT_TRUNC_S(x1, 0);
+
+	return x;
+#else /* XCHAL_HAVE_HIFI3 */
+	return Q_CONVERT_FLOAT(data, 15);
+#endif /* XCHAL_HAVE_HIFI3 */
+}
+
+static inline float convert_int16_to_float(int16_t data)
+{
+#if XCHAL_HAVE_HIFI3
+	const xtfloat ratio = 2 << 15;
+	xtfloat x0 = data;
+	float x;
+
+	x = XT_DIV_S(x0, ratio);
+
+	return x;
+#else /* XCHAL_HAVE_HIFI3 */
+	return Q_CONVERT_QTOF(data, 15);
+#endif /* XCHAL_HAVE_HIFI3 */
+}
+#endif /* CONFIG_COMP_GOOGLE_RTC_USE_32_BIT_FLOAT_API */
+
+/* todo CONFIG_FORMAT_S32LE */
+static int google_rtc_audio_processing_process(struct processing_module *mod,
+					       struct sof_source **sources, int num_of_sources,
+					       struct sof_sink **sinks, int num_of_sinks)
+{
+	int ret;
+	int16_t const *src;
+	int8_t const *src_buf_start;
+	int8_t const *src_buf_end;
+	size_t src_buf_size;
+
+	int16_t const *ref;
+	int8_t const *ref_buf_start;
+	int8_t const *ref_buf_end;
+	size_t ref_buf_size;
+
+	int16_t *dst;
+	int8_t *dst_buf_start;
+	int8_t *dst_buf_end;
+	size_t dst_buf_size;
+
+	size_t num_of_bytes_to_process;
+	size_t channel;
+	size_t buffer_offset;
+
+	struct sof_source *ref_stream, *src_stream;
+	struct sof_sink *dst_stream;
+
+	struct google_rtc_audio_processing_comp_data *cd = module_get_private_data(mod);
+
+	if (cd->reconfigure) {
+		ret = google_rtc_audio_processing_reconfigure(mod);
+		if (ret)
+			return ret;
+	}
+
+	src_stream = sources[cd->raw_microphone_source];
+	ref_stream = sources[cd->aec_reference_source];
+	dst_stream = sinks[0];
+
+	num_of_bytes_to_process = cd->num_frames * source_get_frame_bytes(ref_stream);
+	ret = source_get_data(ref_stream, num_of_bytes_to_process, (const void **)&ref,
```

Comment:
Is it always guaranteed that the input buffers are int16 ? because otherwise the pointer arithmetic to move to the next frame won't work if let's say the input is int32_t.

Reply:
CONFIG_FORMAT_S32LE is not implemented